patina_dxe_core/src/gcd/spin_locked_gcd.rs

```diff
@@ -1146,6 +1152,213 @@ impl GCD {
+            }
+            write_idx += 1;
+        }
+
+        write_idx
+    }
+
+    /// Determines if a GCD memory descriptor should be included in the EFI memory map.
+    ///
+    /// # Arguments
+    /// * `descriptor` - The GCD memory space descriptor
+    ///
+    /// Returns
+    /// * `Some(memory_type)` if the descriptor should be included, `None` otherwise.
+    fn is_efi_memory_map_descriptor(descriptor: &MemorySpaceDescriptor) -> Option<efi::MemoryType> {
+        // Validate page alignment and size
+        let number_of_pages = uefi_size_to_pages!(descriptor.length as usize) as u64;
+        if number_of_pages == 0 {
+            debug_assert!(false, "GCD returned a memory descriptor smaller than a page.");
+            return None; // skip entries for things smaller than a page
+        }
+        if !descriptor.base_address.is_multiple_of(UEFI_PAGE_SIZE as u64) {
+            debug_assert!(false, "GCD returned a non-page-aligned memory descriptor.");
+            return None; // skip entries not page aligned
+        }
+
+        // Apply the same filtering logic as the actual memory map generation
+        memory_type_for_handle(descriptor.image_handle).or({
+            match descriptor.memory_type {
+                // free memory not tracked by any allocator.
+                GcdMemoryType::SystemMemory => Some(efi::CONVENTIONAL_MEMORY),
+
+                // MMIO. Note: there could also be MMIO tracked by the allocators which would not hit this case.
+                GcdMemoryType::MemoryMappedIo => {
+                    // we should only be returning runtime MMIO here
+                    if descriptor.attributes & efi::MEMORY_RUNTIME == 0 { None } else { Some(efi::MEMORY_MAPPED_IO) }
+                }
+
+                // Persistent. Note: this type is not allocatable, but might be created by agents other than the core directly
+                // in the GCD.
+                GcdMemoryType::Persistent => Some(efi::PERSISTENT_MEMORY),
+
+                // Unaccepted. Note: this type is not allocatable, but might be created by agents other than the core directly
+                // in the GCD.
+                GcdMemoryType::Unaccepted => Some(efi::UNACCEPTED_MEMORY_TYPE),
+
+                // Reserved.
+                GcdMemoryType::Reserved => Some(efi::RESERVED_MEMORY_TYPE),
+
+                // Other memory types are ignored for purposes of the memory map
+                _ => None,
+            }
+        })
+    }
+
+    /// Adjusts memory descriptor attributes for the EFI memory map.
+    ///
+    /// ## Arguments
+    ///
+    /// * `descriptor` - The GCD memory space descriptor
+    /// * `memory_type` - The EFI memory type for this descriptor
+    /// * `active_attributes` - If true, use active attributes; if false, use capabilities
+    ///
+    /// Returns
+    /// * `u64` - The adjusted attributes for the EFI memory descriptor.
+    fn adjust_efi_memory_map_descriptor(
```

Comment:
@os-d will probably want this consolidated to his `gcd::MemoryProtectionPolicy` struct

Reply:
Yeah, I remember debating it at the time and I don't remember why I didn't include the runtime setting on runtime code and data to be in the policy. I think having this adjust function be separate from the policy is totally fine, there are two different use cases here and it is really just whether to apply the policy or not.

But at this point, I would say the runtime bit on runtime services code and data should just move into apply_efi_memory_map_policy(), then this function (if it stays a function) can just be if active_attributes, just return attributes, otherwise apply policy.